### DISCUSSION TITLE
Question: Advice on breaking changes for NtsGeometryServices

### DISCUSSION THREAD

**Discussion by m1dst** [NetTopologySuite/NetTopologySuite#495] [Q&A]
Microsoft have some sample code on reprojecting geometry.  This code no longer compiles due to recent updates to NTS.   They have an open issue https://github.com/aspnet/EntityFramework.Docs/issues/1639 for the documentation which can be found at https://github.com/aspnet/EntityFramework.Docs/blob/master/entity-framework/core/modeling/spatial.md

Would you be able to advise how the simple code can be changed to support the recent changes?

**Comment by airbreather** [comment]
Scanning the code sections from top to bottom:

- `IPoint` is now `Point`
- `IGeometry` is now `Geometry`
- The version of ProjNet we maintain no longer depends on GeoAPI, so we've moved from the `ProjNet4GeoAPI` package ID to the `ProjNet` package ID
- `IGeometryServices` is now `NtsGeometryServices`
- `ICoordinateSystemServices` is now `CoordinateSystemServices`
- `IGeometry` is now `Geometry`

**Comment by airbreather** [comment]
Ahh, it uses [`GeometryTransform`](https://github.com/NetTopologySuite/NetTopologySuite/blob/v1.15.3/NetTopologySuite.CoordinateSystems/Transformations/GeometryTransform.cs), which I think isn't there anymore.

This should be equivalent to that method (untested!):
```csharp
using System;

using NetTopologySuite.Geometries;

using ProjNet.CoordinateSystems.Transformations;

public static class GeometryTransform
{
    public static Geometry TransformGeometry(Geometry geometry, MathTransform mathTransform)
    {
        if (geometry is null)
        {
            throw new ArgumentNullException(nameof(geometry));
        }

        if (mathTransform is null)
        {
            throw new ArgumentNullException(nameof(mathTransform));
        }

        geometry = geometry.Copy();
        geometry.Apply(new MathTransformFilter(mathTransform));
        return geometry;
    }

    private sealed class MathTransformFilter : ICoordinateSequenceFilter
    {
        private readonly MathTransform _transform;

        public MathTransformFilter(MathTransform transform) => _transform = transform;

        public bool Done => false;
        public bool GeometryChanged => true;

        public void Filter(CoordinateSequence seq, int i)
        {
            double x = seq.GetX(i);
            double y = seq.GetY(i);
            double z = seq.GetZ(i);
            _transform.Transform(ref x, ref y, ref z);
            seq.SetX(i, x);
            seq.SetY(i, y);
            seq.SetZ(i, z);
        }
    }
}
```

**Comment by FObermaier** [comment]
Note: The matching ProjNet version is not on NuGet yet.

**Comment by airbreather** [comment]
> Note: The matching ProjNet version is not on NuGet yet.

Hmm, I did push a 2.0.0: https://www.nuget.org/packages/ProjNet/2.0.0

**Comment by FObermaier** [comment]
Sorry, it didn't make it to a [Release](https://github.com/NetTopologySuite/ProjNet4GeoAPI/releases) then.

**Comment by bricelam** [comment]
Updated in https://github.com/aspnet/EntityFramework.Docs/pull/1719. It's ugly an inefficient, but I tried to keep the code as compatible with 1.x as I could.

**Comment by IanKemp** [comment]
What is the rationale for removing GeoAPI interfaces and using concrete implementations instead? Doing so makes it very difficult for consumers to test, as they can no longer substitute their own implementations.

BTW you should probably consider updating your wiki...

**Comment by airbreather** [comment]
> What is the rationale for removing GeoAPI interfaces and using concrete implementations instead? Doing so makes it very difficult for consumers to test, as they can no longer substitute their own implementations.

To clarify, basically all the interesting functionality on `NetTopologySuite.Geometries.Geometry` itself is in virtual / abstract methods; compared to older versions, the only extra thing that you should have to do for this type is provide an instance of `GeometryFactory` to the base constructor.
- The out-of-the-box subclasses, such as `NetTopologySuite.Geometries.Polygon`, are less amenable to overriding their specific members, but I don't expect this to block anyone.

Some interesting methods aren't virtual because they're defined in terms of methods that are... e.g., `x.CoveredBy(y)` is defined as `y.Covers(x)`.

The justification stems from the fact that `NetTopologySuite.Geometries.Geometry`, like the counterpart in JTS that it came from, is **already an abstraction**.
- Adding another layer of abstraction on top of it is not just redundant, but because this other layer is an interface, it's harder for us to make non-breaking changes in the future, and it's harder for implementers to take advantage of that abstraction since they have to supply an implementation for **all** members instead of just the `abstract` ones and a subset of the `virtual` ones.
- Default interface implementations aren't coming until C# 8, and the classic .NET Framework will never get them, so we can't add members to the interface to make things easier for consumers without breaking any alternative implementations that might exist.

The discussions we had on the topic can be found across a few issues (plus unrelated discussion, sorry...):
- NetTopologySuite/GeoAPI#30
- NetTopologySuite/GeoAPI#67
- NetTopologySuite/GeoAPI#68
- NetTopologySuite/GeoAPI#70

**Comment by IanKemp** [comment]
OK, that all makes sense. Considering the breadth and breaking-ness of these changes, is it likely we'll see them in the `Microsoft.EntityFrameworkCore.SqlServer.NetTopologySuite` package when that gets bumped to 3.0 with the impending (i.e. literally today) release of .NET Core 3.0?

**Comment by airbreather** [comment]
> is it likely we'll see them in the `Microsoft.EntityFrameworkCore.SqlServer.NetTopologySuite` package when that gets bumped to 3.0 with the impending (i.e. literally today) release of .NET Core 3.0?

Yes; EF Core 3.0's out-of-the-box spatial packages depend on NTS 2.0.

**Comment by airbreather** [comment]
@trevelyanuk please open a new issue for this
